File: 백준/Gold/2573. 빙산/빙산.py
from collections import deque
import sys
input = sys.stdin.readline

dx = [0,1,0,-1]
dy = [1,0,-1,0]
N, M = map(int,input().split())
arr = [list(map(int,input().split())) for _ in range(N)]
# 녹는다
def melt():
    q = deque()
    # 빙산이 남은 경우
    for i in range(N):
        for j in range(M):
            melt_cnt = 0
            if arr[i][j] != 0:
                for k in range(4):
                    nx = i + dx[k]
                    ny = j + dy[k]
                    if 0 <= nx < N and 0 <= ny < M and arr[nx][ny] == 0:
                        melt_cnt +=1
                q.append((i,j,melt_cnt))
    if q:
        while q:
            x, y, cnt = q.popleft()
            if arr[x][y]<cnt:
                arr[x][y] = 0
            else:
                arr[x][y] -= cnt
        return True


    # 빙산이 다 녹았을 경우
    else:
        return False

# 덩어리 몇개인지 구분한다
def check():
    flag = False
    visited = [[False]*M for _ in range(N)]
    for i in range(N):
        for j in range(M):
            if arr[i][j] != 0:
                if not visited[i][j]:
                    BFS(i,j,visited)
                    if flag:
                        return True
                    flag = True
                else:
                    continue
    return False

def BFS(i,j,visited):
    q = deque()
    q.append((i,j))
    visited[i][j] = True
    while q:
        x, y = q.popleft()

        for k in range(4):
            nx = x + dx[k]
            ny = y + dy[k]
            if 0<=nx<N and 0<=ny<M and not visited[nx][ny] and arr[nx][ny]!=0:
                visited[nx][ny] = True
                q.append((nx,ny))

result = 0
while True:
    # 다 녹았으면 종료.
    if not melt():
        result = 0
        break
    result += 1
    # 덩어리가 2개가 되면 종료.
    if check():
        break
print(result)

File: 백준/Gold/2573. 빙산/빙산.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const dir = [
  [0, 1],
  [0, -1],
  [1, 0],
  [-1, 0],
];
const [N, M] = input.shift().trim().split(" ").map(Number);
let arr = input.map((i) => i.trim().split(" ").map(Number));

let time = 0;

function bfs(i, j, visited) {
  let queue = [[i, j]];
  visited[i][j] = true;
  
  while (queue.length > 0) {
    const [x, y] = queue.shift();
    
    for (let [dx, dy] of dir) {
      const nx = x + dx;
      const ny = y + dy;

      if (nx < 0 || nx >= N || ny < 0 || ny >= M) continue;
      if (!visited[nx][ny] && arr[nx][ny] > 0) {
        visited[nx][ny] = true;
        queue.push([nx, ny]);
      }
    }
  }
}

function melt() {
  let next = arr.map((row) => [...row]);
  
  for (let i = 0; i < N; i++) {
    for (let j = 0; j < M; j++) {
      if (arr[i][j] > 0) {
        let seaCount = 0;
        for (let [dx, dy] of dir) {
          const nx = i + dx;
          const ny = j + dy;

          if (nx < 0 || nx >= N || ny < 0 || ny >= M) continue;
          if (arr[nx][ny] === 0) seaCount++;
        }
        next[i][j] = Math.max(0, arr[i][j] - seaCount);
      }
    }
  }
  arr = next;
}

while (true) {
  let icebergCount = 0;
  let visited = Array.from({ length: N }, () => Array(M).fill(false));
  
  for (let i = 0; i < N; i++) {
    for (let j = 0; j < M; j++) {
      if (arr[i][j] > 0 && !visited[i][j]) {
        icebergCount++;
        bfs(i, j, visited);
      }
    }
  }

  if (icebergCount === 0) {
    console.log(0);
    break;
  }
  
  if (icebergCount >= 2) {
    console.log(time);
    break;
  }

  melt();
  time++;
}
